### ISSUE TITLE
Option to disallow opening "real" SD card path?

### ISSUE THREAD

**Issue by AdityaAnand1** [spacecowboy/NoNonsense-FilePicker#158]
AFAIK, there's three "kinds" of storage:

* Internal: Where we can store our app's internal stuff, cache etc.
* "External": The "normal" storage, which is still just the internal SD card partitioned into two (afaik)
* The actually "external"/user-inserted SD card: That this library (or really by any means) we can't really use because Google.

What I've found is that I CAN write my files to the actually external SD card if it is within `/Android/media/<my-package>`. 

Now comes the question of what behavior would be best for this app, a file picker, with an optional ability to create folders.

I don't want to show the users the top "...", so that they can climb out of the fake-"external" SD card and go into the "real" external SD card where they CANNOT write anything except within `/Android/media/<my-package>`. 

So, a user picking out a path in the "real" external storage (which is possible because we can read, but not write to it) is of no use to me because the path that I get back is one that I cannot write to and will have to explain that to the user (and they still won't get it and come and 1 star me). The only path that we can do anything with is `/Android/media/<my-package>`.

So, I think the best behavior would be to just let the library be as it is for the external storage and simply not let the users climb into the "real" sd card. This way, I can manage my own direction (`/Android/media/<my-package>`) myself and use the file picker where it does work normally.

All that is needed for the above to work is perhaps an option to disallow letting users climb out of the fake "external" SD card and into the "real" one. An option to not show "..." at the root fake-
 `ExternalStorageDirectory()`.

Does that make sense?

**Comment by spacecowboy** [comment]
Indeed. Google have been quite un-helpful here.

You have two approaches.

One option would be to specify the [RootPath](https://github.com/spacecowboy/NoNonsense-FilePicker/blob/e4167c3aab0c1c161de634c34fdff30cd4763d69/library/src/main/java/com/nononsenseapps/filepicker/LogicHandler.java#L76) to prevent a user from navigating out of the internal data directory. This is not fool proof however. You'd really need some additional logic to make sure a directory is not on an external sd card. But that's really error prone because it's not obvious that it's on an external sd card.

The better approach would be to use Android's built-in picker for Kitkat+ devices. Example code ([from Feeder](https://gitlab.com/spacecowboy/Feeder/blob/master/app/src/main/kotlin/com/nononsenseapps/feeder/ui/FeedActivity.kt)):

```kotlin
fun openPicker() {
  val intent: Intent
  if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
    intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
    intent.type = "text/opml"
    intent.putExtra(Intent.EXTRA_TITLE, "feeder.opml")
  } else {
    intent = Intent(this, MyFilePickerActivity::class.java)
    intent.putExtra(AbstractFilePickerActivity.EXTRA_MODE, AbstractFilePickerActivity.MODE_NEW_FILE)
    intent.putExtra(AbstractFilePickerActivity.EXTRA_ALLOW_EXISTING_FILE, true)
    intent.putExtra(AbstractFilePickerActivity.EXTRA_START_PATH,  File(Environment.getExternalStorageDirectory(), "feeder.opml").path)
  }
  startActivityForResult(intent, EXPORT_OPML_CODE)
}

override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
  if (data != null) {
    val uri = data.data
    try {
      doSomethingWithFile(uri)
    } catch (e: Throwable) {
      e.printStackTrace()
    }
  }
}
```

**Comment by AdityaAnand1** [comment]
Oh no.

>The better approach would be to use Android's built-in picker for Kitkat+

I hadn't looked at the [statistics](https://developer.android.com/about/dashboards/index.html) in quite a while. It doesn't make much sense to support <Kitkat now (<5% market share)

@spacecowboy I'd just like to thank you for all the work you put in this library. I remember when it took off, around v1.2 I think, back in 2014 (I think?). Google was, as usual, of no help, to a budding android developer like me, looking for an elegant solution, to what I imagine is one of the most simplest tasks a platform should natively do well, picking a file/folder. 

Then one day I found your library. It was... perfection. Simple API? Check. Elegant UX? Check. It even had material design before material design really took off. I've used this in pretty much every project I've ever worked on. It has saved me from countless hours of googling and fighting with the native API.

My users will have a hard time moving on. Hundreds of thousands of users of [my apps](https://play.google.com/store/apps/developer?id=Xeus) are now used to this simple and familiar file picker. I expect dozens of one stars and thousands of users wondering why I would change the file picker. But I must. Offering two file pickers will be more confusing. Explaining why I can't write to the removable SD card with this file picker will be impossible.

So now I must remove this library from all my projects. @spacecowboy Thank you for all your hard work. It has been an honor and a pleasure to use your library. I wish you all the very best in whatever you do.

**Comment by spacecowboy** [comment]
👍 

Thank you for your kind words @AdityaAnand1 . It means a lot. Also very cool to see the library being used so much.

I've added a [note](https://github.com/spacecowboy/NoNonsense-FilePicker#note-avoid-using-as-sd-card-file-picker-on-kitkat) to the README to help other avoid this particular issue.

**Comment by AdityaAnand1** [comment]
Actually, scratch that. This library just goes too well with my app and I'm not giving it up without a fight. And I just found out that you can't even send the SAF directory picker to a custom starting directory like you can with this library (before [freaking API 26](https://developer.android.com/reference/android/provider/DocumentsContract.html#EXTRA_INITIAL_URI), wtf Google). That's just too much. My users will hate me if I switch and actually take away functionality.

I think I found [a way](https://stackoverflow.com/questions/36862675/android-sd-card-write-permission-using-saf-storage-access-framework) to ask permission for external SD card, the way that file managers like ES or Amaze do it. Ask the user to give the "removable SD card root directory" permission and then just work with files normally.

Have you explored this option? I think it might even work with this library. A simple utility method to send the user off to get back permission for the removable SD card from SAF picker whenever the user picks a directory that we can't write to (which we would be able to detect).

I'm not sure if after getting that "permission" we can use the simple java `File` API or do we still have to go through `DocumentFile`.

**Comment by spacecowboy** [comment]
It's certainly worth looking into. If a permission can be granted via the ContentResolver API directly without popping up the SAF picker then we're in business. I'll write some experimental code to investigate.

**Comment by spacecowboy** [comment]
Did some testing and it seems there is no way to avoid opening the SAF picker in order to get URI permission. Not using it and trying to get a permission results in

```
java.lang.SecurityException: UID 10078 does not have permission to content://com.android.externalstorage.documents/tree/%2Fstorage%2Femulated%2F0 [user 0]; you could obtain access using ACTION_OPEN_DOCUMENT or related APIs
```

It's quite possible that I haven't explored all possibilities but it looks like you can't get around the requirement of SAF on Kitkat+.

**Comment by AdityaAnand1** [comment]
Yeah, the user's physical act of picking a directory in the SAF is what grants us the permission to the directory tree. No way around it.

I'm quite sure there won't be any workarounds to having to open the SAF picker as that would undermine it's very existence. So what I was thinking was the next best thing we can do. Now this would be only for devs (like me) that prefer their custom file picker (like this library or custom implementations in case of file explorers) over the SAF picker.

* Explain to the user, maybe put this in an app's onboarding, that they need to pick the root directory in the SAF picker for the app to have permission to write files on the ext SD card.
* The way you do that would be something like: Open up the dialog (could be bundled in the library, maybe through a util method for those that want SD card writing) explaining the above (like all the file managers today are doing, like Amaze or ES or whatever). Start an activity that opens up the SAF picker and in it's `onActivityResult` listens for the SAF callback getting us the DocumentTree, the permission for which we can then persist for future use. This would be the same dialog as in [this stack overflow question](https://stackoverflow.com/questions/36862675/android-sd-card-write-permission-using-saf-storage-access-framework).
* When you get back the permission to work with the ext SD card (a one time only task), you then get the ability to use your custom file picker since you have permission to work with ext SD. What I'm not sure about is would this new permission allow us to work with the plain `File` API or do we have to put everything in an outputStream and pipe it through the new SAF `DocumentFile` API for creating new files. Either way, doing the above seems to be pretty much the only option.

The thing is, API 26 has fixed a lot of things wrong with SAF, but they can't put it in the support library since it's platform behavior. So using SAF for those who have advanced use cases (again, like me and like file managers) will suck for at least another 3-4 years until our minSdkVersion becomes 26 (Thanks Google). That's not much of a solution if you ask me, but that's what we got. So implementing this, if it does work, would easily double this library's lifetime.

Actually, now that I think of it, the above doesn't really belong in the library. There's not that much to it and not everybody would need it but what we could have is a simple page in the wiki detailing how to do the above (open dialog, then go to SAF, then listen for response) themselves before the library works correctly for ext SD card. I'm actually implementing this in my own app right now, so I'll find out soon enough whether this method works or not.

**Comment by AdityaAnand1** [comment]
After extensive testing and digging around in AOSP and open source file managers (like Amaze), there doesn't seem to be a way around this.

The `File` API should now be considered deprecated for all use cases except perhaps managing an app's private directories. There is no way to write a `File` anymore. You can only write a `DocumentFile`.

**Comment by spacecowboy** [comment]
Too bad :(

**Comment by przemeknzb** [comment]
@AdityaAnand1 your solution will work, here is a util class that lets you use the standard File apis.  It even works on the latest Android P.
https://github.com/jeisfeld/Augendiagnose/blob/5f7478f5847a76dda28f79721d67df0b340f97a3/AugendiagnoseIdea/augendiagnoseLib/src/main/java/de/jeisfeld/augendiagnoselib/util/imagefile/FileUtil.java

**Comment by AdityaAnand1** [comment]
@przemeknzb Looks interesting! Will try it out and report back 😄
